tools/sync/commands.py

```diff
@@ -67,35 +70,68 @@ def __init__(self, base):
         self.logger = logging.getLogger(__name__)
         logging.basicConfig()
 
+    def run_command(command):
+        cmd = Command(command,
+                      args_subst={"PROJECT": self.name},
+                      args_append=[self.name], excl_subst=True)
+        cmd.execute()
+        self.retcodes[str(cmd)] = cmd.getretcode()
+        self.outputs[str(cmd)] = cmd.getoutput()
+
     def run(self):
         """
         Run the sequence of commands and capture their output and return code.
         First command that returns code other than 0 terminates the sequence.
         If the command has return code 2, the sequence will be terminated
         however it will not be treated as error.
+
+        Any command entry that is a URI, will be used to submit RESTful API
+        request. Return codes for these requests are not checked.
         """
 
+        PROJECT_SUBST = '%PROJECT%'
         for command in self.commands:
-            cmd = Command(command,
-                          args_subst={"ARG": self.name},
-                          args_append=[self.name], excl_subst=True)
-            cmd.execute()
-            self.retcodes[str(cmd)] = cmd.getretcode()
-            self.outputs[str(cmd)] = cmd.getoutput()
-
-            # If a command fails, terminate the sequence of commands.
-            retcode = cmd.getretcode()
-            if retcode != 0:
-                if retcode == 2:
-                    self.logger.info("command '{}' requested break".
-                                     format(cmd))
-                    self.run_cleanup()
+            if is_web_uri(command[0]):
+                uri = command[0].replace(PROJECT_SUBST, self.name)
+                verb = command[1]
+                data = command[2]
+                if len(data) == 0:
+                    # PUT/POST need the data.
+                    if verb in ['PUT', 'POST']:
+                        self.logger.error('Empty data for PUT/POST')
+                        continue
+                    data = None
+                else:
+                    headers = {'Content-Type': 'application/json'}
+                    json_data = json.dumps(data).replace(PROJECT_SUBST,
+                                                         self.name)
+                    self.logger.debug("JSON data: {}".format(json_data))
+
+                if verb == 'PUT':
+                    put(self.logger, uri, headers=headers, data=json_data)
```

Comment:
`requests` have `json` param which will do conversion to json for you. Not sure if this is what you want since you are doing string replacement there... Just a remark :)

Reply:
However, if you use it then it won't be needed to use custom headers yet. (I hope it sets the `application/json` header by itself).

Reply:
I know about the `json` param. However, the project needs to be substituted first. Also, the version of `requests` I am working with does not have the param yet.

Reply:
I looked in the doc and it is there http://docs.python-requests.org/en/v1.0.4/user/quickstart/ and the 1.0.4 is from the 2012, which kind of version do you use?